tracer/src/instruction/test.rs

```diff
@@ -1,89 +1,246 @@
-use rand::{rngs::StdRng, RngCore, SeedableRng};
+//! Virtual instruction‑sequence fuzzer split into **non‑memory** and **memory** cases.
+//!
+//! The harness executes the real instruction on one CPU and its virtual
+//! sequence on another, then compares registers.
+//! The expectation is that virtual sequences have the same behavior as a real instruction
+//! on real registers
+use std::any::TypeId;
 
+use super::format::format_load::FormatLoad;
+use super::format::format_r::FormatR;
+use super::format::format_s::FormatS;
+use super::{
+    div::DIV, divu::DIVU, lb::LB, lbu::LBU, lh::LH, lhu::LHU, mulh::MULH, mulhsu::MULHSU, rem::REM,
+    remu::REMU, sb::SB, sh::SH, sll::SLL, slli::SLLI, sra::SRA, srai::SRAI, srl::SRL, srli::SRLI,
+    RISCVInstruction, RISCVTrace, VirtualInstructionSequence,
+};
+use super::{RISCVCycle, RV32IMCycle};
 use crate::emulator::cpu::{Cpu, Xlen};
+use crate::emulator::terminal::DummyTerminal;
+use rand::{rngs::StdRng, RngCore, SeedableRng};
+
+const DRAM_BASE: u64 = 0x8000_0000;
+const DRAM_SIZE: u64 = 1024;
+
+// Implements a `from_operands` constructor for store instructions
+#[cfg(test)]
+macro_rules! impl_from_operands_s {
+    ($($ty:ident),* $(,)?) => {
+        $(
+            #[cfg(test)]
+            impl $ty {
+                pub fn from_operands(rs1: usize, rs2: usize, imm: i64, address: u64) -> Self {
```

Comment:
If you use `InstructionFormat::random()` and `InstructionRegisterState::random()`, as well as `normalize()` on instruction, these from_operands implementations can be removed.